src/main/java/io/cdap/plugin/couchbase/source/CouchbaseSource.java

```diff
@@ -118,4 +137,98 @@ public void transform(KeyValue<NullWritable, N1qlQueryRow> input, Emitter<Struct
+      cluster.authenticate(config.getUser(), config.getPassword());
+    }
+    String bucketName = config.getBucket();
+    Bucket bucket = cluster.openBucket(bucketName);
+    String inferStatement = String.format("INFER `%s` WITH {\"sample_size\":%d};", bucketName, config.getSampleSize());
+    N1qlQuery query = N1qlQuery.simple(inferStatement);
+    N1qlQueryResult result = bucket.query(query);
+    if (!result.finalSuccess()) {
+      String errorMessage = result.errors().stream()
+        .map(JsonObject::toString)
+        .collect(Collectors.joining("\n"));
+      throw new InvalidStageException(String.format("Unable to infer output schema: '%s'", errorMessage));
+    }
+
+    // row.value() can not be used since it expects the result to be a JSON document. However, the result is an array
+    // that can not be deserialized as JsonObject
+    N1qlQueryRow row = result.rows().next();
+    String json = row.toString();
+    // Deserialize the output in the JSON Schema draft v4 format as JsonArray
+    // The output is an array of schema documents, each of them groups properties for docs with similar structure
+    // See: https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/infer.html#example
+    JsonArray couchbaseSchema = JsonArray.fromJson(json);
+
+    List<Schema.Field> fields = new ArrayList<>();
+    // Select fields may contain metadata fields such as meta(`travel-sample`).id
+    // Include them to the inferred schema as well using simple names after dot character
+    // All metadata fields are strings
+    if (config.getSelectFieldsList() != null) {
+      List<Schema.Field> metadataFields = config.getSelectFieldsList().stream()
+        .filter(f -> f.startsWith("meta"))
+        .map(f -> f.substring(f.lastIndexOf(".") + 1))
+        .map(simpleName -> Schema.Field.of(simpleName, Schema.of(Schema.Type.STRING)))
+        .collect(Collectors.toList());
+      fields.addAll(metadataFields);
+    }
+
+    for (int i = 0; i < couchbaseSchema.size(); i++) {
+      JsonObject schema = couchbaseSchema.getObject(i);
+      List<Schema.Field> schemaFields = recordSchema(schema, config.getSelectFieldsList());
+
+      // Schemas ordered by number of sample documents.
+      // Use inferred type that matches most of the documents
+      schemaFields.stream()
+        .filter(f -> !fields.contains(f))
+        .forEach(fields::add);
+    }
+
+    return Schema.recordOf("inferred-schema", fields);
+  }
+
+  private List<Schema.Field> recordSchema(JsonObject recordMetadata, @Nullable List<String> selectFields) {
+    JsonObject couchbasePropertiesMetadata = recordMetadata.getObject("properties");
+
+    List<Schema.Field> fields = new ArrayList<>();
+    for (String propertyName : couchbasePropertiesMetadata.getNames()) {
+      if (selectFields != null && !selectFields.contains(propertyName) && !selectFields.contains("*")) {
```

Comment:
it does not make sense to do this loop if selectFields == null or is empty

Reply:
The `selectFields` used only for the top-level record.
Couchbase properties of type `object` are mapped to the record using 
`recordSchema(JsonObject recordMetadata, @Nullable List<String> selectFields)` 
method:
```
...
case "object":
    List<Schema.Field> objectFields = recordSchema(propertyMetadata, null);
    return Schema.nullableOf(Schema.recordOf(propertyName + "-inferred-nested", objectFields));
...
```